File: chatbot-directline-api/php_demo/step4_get_messages.php
<?php
/**
 * Step 4: Get conversation messages and poll for replies
 */

require_once 'config.php';

function getMessages($watermark = null) {
    if ($watermark) {
        echo "Step 4: Checking for new messages since watermark: {$watermark}\n";
    } else {
        echo "Step 4: Getting all conversation messages...\n";
    }
    
    // Ensure data directory exists
    $dataDir = 'data';
    if (!is_dir($dataDir)) {
        mkdir($dataDir, 0755, true);
    }
    
    // Read token and conversation ID from previous steps
    $tokenFile = $dataDir . '/token.txt';
    $conversationIdFile = $dataDir . '/conversation_id.txt';
    
    if (!file_exists($tokenFile) || !file_exists($conversationIdFile)) {
        echo "❌ Token or conversation ID not found. Run previous steps first.\n";
        return null;
    }
    
    $token = trim(file_get_contents($tokenFile));
    $conversationId = trim(file_get_contents($conversationIdFile));
    
    $url = "https://directline.botframework.com/v3/directline/conversations/{$conversationId}/activities";
    if ($watermark) {
        $url .= "?watermark={$watermark}";
    }
    
    $headers = ['Authorization: Bearer ' . $token];
    
    // Prepare cURL request
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
    
    // Get messages
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    
    if ($httpCode === 200) {
        $data = json_decode($response, true);
        $activities = $data['activities'] ?? [];
        $newWatermark = $data['watermark'] ?? null;
        
        echo "✅ Retrieved " . count($activities) . " message(s)\n";
        echo "   New watermark: {$newWatermark}\n";
        
        // Only display messages if not using watermark (i.e., getting all messages)
        if (!$watermark) {
            // Display messages
            foreach ($activities as $activity) {
                $sender = $activity['from']['id'] ?? 'unknown';
                $text = $activity['text'] ?? '';
                $timestamp = $activity['timestamp'] ?? '';
                
                echo "   📨 {$sender}: {$text}\n";
                echo "      Time: {$timestamp}\n";
            }
        }
        
        // Save watermark for next check
        file_put_contents($dataDir . '/watermark.txt', $newWatermark);
        
        return $data;
    }
    
    echo "❌ Failed to get messages: HTTP {$httpCode}\n";
    if ($response) {
        echo "Response: {$response}\n";
    }
    
    return null;
}

function getCurrentWatermark() {
    // Ensure data directory exists
    $dataDir = 'data';
    if (!is_dir($dataDir)) {
        mkdir($dataDir, 0755, true);
    }
    
    // Read token and conversation ID from previous steps
    $tokenFile = $dataDir . '/token.txt';
    $conversationIdFile = $dataDir . '/conversation_id.txt';
    
    if (!file_exists($tokenFile) || !file_exists($conversationIdFile)) {
        echo "❌ Token or conversation ID not found. Run previous steps first.\n";
        return null;
    }
    
    $token = trim(file_get_contents($tokenFile));
    $conversationId = trim(file_get_contents($conversationIdFile));
    
    $url = "https://directline.botframework.com/v3/directline/conversations/{$conversationId}/activities";
    $headers = ['Authorization: Bearer ' . $token];
    
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
    
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    
    if ($httpCode === 200) {
        $data = json_decode($response, true);
        return $data['watermark'] ?? null;
    }
    
    echo "❌ Failed to get current watermark: HTTP {$httpCode}\n";
    return null;
}

function pollForReply($maxAttempts = 10, $interval = 2) {
    echo "\n🔄 Polling for bot reply (max {$maxAttempts} attempts, {$interval}s interval)...\n";
    
    // Get current watermark before polling
    $watermark = getCurrentWatermark();
    if ($watermark === null) {
        return null;
    }
    
    echo "📌 Starting to poll from watermark: {$watermark}\n";
    
    // Poll for new messages
    for ($attempt = 1; $attempt <= $maxAttempts; $attempt++) {
        echo "\nAttempt {$attempt}/{$maxAttempts}...\n";
        sleep($interval);
        
        $data = getMessages($watermark);
        if (!$data) {
            continue;
        }
        
        $activities = $data['activities'] ?? [];
        $newWatermark = $data['watermark'] ?? null;
        
        // Check if we got new messages
        if (count($activities) > 0) {
            echo "🎉 Bot replied! Found " . count($activities) . " new message(s)\n";
            
            // Display the new messages
            foreach ($activities as $activity) {
                $sender = $activity['from']['id'] ?? 'unknown';
                $text = $activity['text'] ?? '';
                $timestamp = $activity['timestamp'] ?? '';
                
                echo "   📨 {$sender}: {$text}\n";
                echo "      Time: {$timestamp}\n";
            }
            
            return $data;
        }
        
        $watermark = $newWatermark;
    }
    
    echo "⏰ No reply received within time limit\n";
    return null;
}

// Run if called directly
if (basename(__FILE__) === basename($_SERVER['SCRIPT_NAME'])) {
    // First get all messages
    getMessages();
    
    // Then poll for new reply
    pollForReply();
}
?> 

File: chatbot-directline-api/python_demo/step4_get_messages.py
import requests
import time
import os


def get_messages(watermark=None):
    """Step 4: Get conversation messages"""
    if watermark:
        print(f"Step 4: Checking for new messages since watermark: {watermark}")
    else:
        print("Step 4: Getting all conversation messages...")

    # Get the directory where this script is located
    script_dir = os.path.dirname(os.path.abspath(__file__))
    data_dir = os.path.join(script_dir, "data")
    os.makedirs(data_dir, exist_ok=True)

    # Read token and conversation ID from previous steps
    try:
        with open(os.path.join(data_dir, "token.txt"), "r") as f:
            token = f.read().strip()
        with open(os.path.join(data_dir, "conversation_id.txt"), "r") as f:
            conversation_id = f.read().strip()
    except FileNotFoundError:
        print("❌ Token or conversation ID not found. Run previous steps first.")
        return None

    url = f"https://directline.botframework.com/v3/directline/conversations/{conversation_id}/activities"
    if watermark:
        url += f"?watermark={watermark}"

    headers = {"Authorization": f"Bearer {token}"}

    # Get messages
    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        data = response.json()
        activities = data.get("activities", [])
        new_watermark = data.get("watermark")

        print(f"✅ Retrieved {len(activities)} message(s)")
        print(f"   New watermark: {new_watermark}")

        # Only display messages if not using watermark (i.e., getting all messages)
        # When using watermark, we only want to know if there are new messages, not display them
        if not watermark:
            # Display messages
            for activity in activities:
                sender = activity.get("from", {}).get("id", "unknown")
                text = activity.get("text", "")
                timestamp = activity.get("timestamp", "")

                print(f"   📨 {sender}: {text}")
                print(f"      Time: {timestamp}")

        # Save watermark for next check
        with open(os.path.join(data_dir, "watermark.txt"), "w") as f:
            f.write(str(new_watermark))

        return data
    else:
        print(f"❌ Failed to get messages: {response.text}")
        return None


def get_current_watermark():
    """Get current watermark without displaying messages"""
    # Get the directory where this script is located
    script_dir = os.path.dirname(os.path.abspath(__file__))
    data_dir = os.path.join(script_dir, "data")
    os.makedirs(data_dir, exist_ok=True)

    try:
        with open(os.path.join(data_dir, "token.txt"), "r") as f:
            token = f.read().strip()
        with open(os.path.join(data_dir, "conversation_id.txt"), "r") as f:
            conversation_id = f.read().strip()
    except FileNotFoundError:
        print("❌ Token or conversation ID not found. Run previous steps first.")
        return None

    url = f"https://directline.botframework.com/v3/directline/conversations/{conversation_id}/activities"
    headers = {"Authorization": f"Bearer {token}"}

    response = requests.get(url, headers=headers)
    if response.status_code == 200:
        data = response.json()
        return data.get("watermark")
    else:
        print(f"❌ Failed to get current watermark: {response.text}")
        return None


def poll_for_reply(max_attempts=10, interval=2):
    """Poll for bot reply using watermark"""
    print(
        f"\n🔄 Polling for bot reply (max {max_attempts} attempts, {interval}s interval)..."
    )

    # Get current watermark before polling
    watermark = get_current_watermark()
    if watermark is None:
        return None

    print(f"📌 Starting to poll from watermark: {watermark}")

    # Poll for new messages
    for attempt in range(max_attempts):
        print(f"\nAttempt {attempt + 1}/{max_attempts}...")
        time.sleep(interval)

        data = get_messages(watermark)
        if not data:
            continue

        activities = data.get("activities", [])
        new_watermark = data.get("watermark")

        # Check if we got new messages
        if len(activities) > 0:
            print(f"🎉 Bot replied! Found {len(activities)} new message(s)")
            # Display the new messages
            for activity in activities:
                sender = activity.get("from", {}).get("id", "unknown")
                text = activity.get("text", "")
                timestamp = activity.get("timestamp", "")

                print(f"   📨 {sender}: {text}")
                print(f"      Time: {timestamp}")
            return data

        watermark = new_watermark

    print("⏰ No reply received within time limit")
    return None


if __name__ == "__main__":
    # First get all messages
    get_messages()

    # Then poll for new reply
    poll_for_reply()
